File: python/herencia/cuentas_bancarias/banco.py
"""Implementa la clase Banco que contiene una colección de Cuentas, incluye los métodos:
abrirCuentaCorriente(dni, descubierto)
abrirCajaDeAhorro(int dni)
totalSaldoEnDescubierto() -> devuelve la suma de los saldos de todas las cuentas corrientes
que están en descubierto.
listarCuentas() -> que muestra los datos de todas las cuentas del banco,
ordenado por saldo de manera ascendente.
"""

from cuentas.cuenta_ahorro import CajaDeAhorro
from cuentas.cuenta_corriente import CuentaCorriente


class Banco:
    """Class Banco"""

    def __init__(self) -> None:
        self.__cuentas = []

    def abrir_caja_ahorro(self, dni) -> CajaDeAhorro:
        """Crea una caja de ahorro y la agrega a la lista de cuentas"""
        return CajaDeAhorro(dni)

    def abrir_cuenta_corriente(self, dni, descubierto) -> CuentaCorriente:
        """Crea una cuenta corriente y la agrega a la lista de cuentas"""
        return CuentaCorriente(dni, descubierto)

    def agregar_cuenta(self, c):
        """Agrega la cuenta c a la lista de cuentas del banco"""
        self.__cuentas.append(c)

    def listar_cuentas(self):
        """Muestra por pantalla la lista de cuentas del banco"""
        for c in self.__cuentas:
            print(c)

    def listar_cuentas_en_descubierto(self):
        """Muestra por pantalla la lista de cuentas corrientes que estan
        en descubierto"""
        for c in self.__cuentas:
            if isinstance(c, CuentaCorriente):
                if c.get_saldo() < 0:
                    print(c)


bco = Banco()
cc1 = bco.abrir_cuenta_corriente(12345678, 10000)
ca1 = bco.abrir_caja_ahorro(23456789)
bco.agregar_cuenta(cc1)
bco.agregar_cuenta(ca1)
bco.listar_cuentas()
cc1.depositar((50000))
ca1.depositar(20000)
ca1.reservar(5000)
print("Listado de cuentas: ")
bco.listar_cuentas()
cc1.extraer(55000)
print("Listado de cuentas en descubierto: ")
bco.listar_cuentas_en_descubierto()

File: java/CuentasBancarias/src/cuentasBancarias/Banco.java
package cuentasBancarias;

import java.util.ArrayList;

public class Banco {

    ArrayList<Cuenta> cuentas;

    public Banco() {
        this.cuentas = new ArrayList<Cuenta>();
    }

    public CuentaCorriente abrirCuentaCorriente(int dni, double descubierto) {
        return new CuentaCorriente(dni, descubierto);
    }

    public CajaAhorro abrirCajaDeAhorro(int dni) {
        return new CajaAhorro(dni);
    }

    public void agregarCuenta(Cuenta c) {
        this.cuentas.add(c);
    }

    public void listarCuentas() {
        for (Cuenta cadaCuenta : this.cuentas) {
            System.out.println(cadaCuenta);
        }
    }

    public static void main(String[] args) {
        Banco coban = new Banco();

        CajaAhorro ca1 = coban.abrirCajaDeAhorro(12345678);
        CuentaCorriente cc1 = coban.abrirCuentaCorriente(3456123, 25000);
        CajaAhorro ca2 = coban.abrirCajaDeAhorro(5671245);

        coban.agregarCuenta(ca2);
        coban.agregarCuenta(cc1);
        coban.agregarCuenta(ca1);

        coban.listarCuentas();

        ca1.depositar(100000);
        ca1.reservar(20000);
        cc1.depositar(50000);
        cc1.extraer(34567.67);
        cc1.transferir(ca2, 60000);
        System.out.println("-------------------------------------");
        coban.listarCuentas();

    }

}
